dbt/include/spark/macros/materializations/table.sql

```diff
@@ -41,7 +41,40 @@
 # --- Autogenerated dbt materialization code. --- #
 dbt = dbtObj(spark.table)
 df = model(dbt, spark)
-df.write.mode("overwrite").format("delta").saveAsTable("{{ target_relation }}")
+
+import importlib.util
+
+pandas_available = False
+pyspark_available = False
+
+# make sure pandas exists before using it
+if importlib.util.find_spec("pandas"):
+  import pandas
+  pandas_available = True
+
+# make sure pyspark.pandas exists before using it
+if importlib.util.find_spec("pyspark.pandas"):
+  import pyspark.pandas
+  pyspark_available = True
+
+# preferentially convert pandas DataFrames to pandas-on-Spark DataFrames first
+# since they know how to convert pandas DataFrames better than `spark.createDataFrame(df)`
+# and converting from pandas-on-Spark to Spark DataFrame has no overhead
+if pyspark_available and pandas_available and isinstance(df, pandas.core.frame.DataFrame):
+  df = pyspark.pandas.frame.DataFrame(df)
```

Comment:
At the risk of making this even more complex than it needs to be — I believe `pyspark.pandas` was introduced in Spark v3.2, and won't be available in earlier versions. (The same functionality was available via the `koalas` package, which was the old codename for pandas-on-PySpark)

Reply:
I _think_ what will happen currently if the input is a pandas-on-Spark DataFrame but `pyspark.pandas` is not available:
-   `msg = f"{type(df)} is not a supported type for dbt Python materialization"`

What I believe will happen if the input is a pandas DataFrame but `pyspark.pandas` is not available:
-   `df = spark.createDataFrame(df)`

We can add in an attempt to `import databricks.koalas` so we are covering as many bases as possible. If we go that route, is there an environment we could test it out on?

Reply:
> If we go that route, is there an environment we could test it out on?

We could spin up a Databricks cluster running an older Spark version (v3.1). This is also what will be running inside Dataproc — the latest Apache Spark release it supports is v3.1.